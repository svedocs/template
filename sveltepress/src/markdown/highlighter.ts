import type { Highlighter } from '@sveltepress/vite'
import type { BundledTheme, CodeToHastOptions, HighlighterGeneric, ShikiTransformer } from 'shiki'
import type { BundledLanguage } from 'shiki/langs'
import type { DefaultThemeOptions } from 'virtual:sveltepress/theme-default'
import { env } from 'node:process'
import { createTransformerFactory } from '@shikijs/twoslash'
import { createTwoslasher, rendererFloatingSvelte } from '@sveltepress/twoslash'
import { LRUCache } from 'lru-cache'
import { createHighlighter } from 'shiki'
import { themeOptionsRef } from '../index.js'
import { processCommands } from './commands.js'

const DEFAULT_SUPPORT_LANGUAGES: any[] = ['svelte', 'sh', 'js', 'html', 'ts', 'md', 'css', 'scss']

const cache = new LRUCache<string, any>({ max: 200 })

const shikiHighlighterInstance: {
  value: null | HighlighterGeneric<BundledLanguage, BundledTheme>
  twoslashTransformer?: ShikiTransformer
} = {
  value: null,
}

const _highlighter: Highlighter = async (code, lang) => {
  const highlighterConfig = themeOptionsRef.value?.highlighter
  const darkTheme = highlighterConfig?.themeDark ?? 'night-owl'
  const lightTheme = highlighterConfig?.themeLight ?? 'vitesse-light'

  const shikiOptions: CodeToHastOptions<BundledLanguage, BundledTheme> = {
    lang,
    themes: {
      dark: darkTheme,
      light: lightTheme,
    },
    transformers: [],
  }
  if (shikiHighlighterInstance.twoslashTransformer) {
    shikiOptions.transformers?.push(
      shikiHighlighterInstance.twoslashTransformer,
    )
  }
  const { codeToHtml } = shikiHighlighterInstance.value!
  return codeToHtml(code, shikiOptions)
    .replace(/\{/g, '&#123;')
    .replace(/\}/g, '&#125;')
    .replace(/<!--svp-floating-snippet-start-->/g, '{#snippet floatingContent()}')
    .replace(/<!--svp-floating-snippet-end-->/g, '{/snippet}')
}

const highlighter: Highlighter = async (code, lang, meta) => {
  const cacheKey = JSON.stringify({ code, lang, meta })
  let cached
  if (env.NODE_ENV === 'development') {
    cached = cache.get(cacheKey)
    if (cached)
      return cached
  }
  const metaArray = (meta || '').split(' ')
  const containLineNumbers = metaArray.some(item => item.trim() === 'ln')
  const titleMeta = metaArray.find(item => item.startsWith('title='))
  const commandDoms: string[] = []
  const lines = code.split('\n')
  let noErrorsFirstLine: string | undefined
  if (lines[0] === '// @noErrors')
    noErrorsFirstLine = lines.shift()
  code = lines.map((line, i) => {
    const [commandDomsInOneLine, newLine] = processCommands(line, i, lines.length)
    commandDoms.push(...commandDomsInOneLine)
    return newLine
  }).join('\n')
  let title: string | undefined
  if (titleMeta)
    title = titleMeta.split('=')[1].replace(/(^")|("$)/g, '')
  if (noErrorsFirstLine)
    code = `${noErrorsFirstLine}\n${code}`

  cached = `
<div class="svp-code-block-wrapper">${title
  ? `<div class="svp-code-block--title">${title}</div>
`
  : ''}
  <div class="svp-code-block${containLineNumbers ? ' svp-code-block--with-line-numbers' : ''}">
    ${commandDoms.join('\n')}
    <!-- svelte-ignore a11y_no_noninteractive_tabindex -->
    ${await _highlighter(code, lang, meta)}
    <div class="svp-code-block--lang">
      ${lang}
    </div>
    <CopyCode />
    ${containLineNumbers ? `<div class="svp-code-block--line-numbers">${lines.map((_, i) => `<div class="svp-code-block--line-number-item">${i + 1}</div>`).join('\n')}</div>` : ''}
  </div>
</div>`
  if (env.NODE_ENV === 'development')
    cache.set(cacheKey, cached)
  return cached
}

export async function initHighlighter(highlighterConfig: DefaultThemeOptions['highlighter'] = {}) {
  const langs = highlighterConfig?.languages || DEFAULT_SUPPORT_LANGUAGES
  const darkTheme = highlighterConfig?.themeDark ?? 'night-owl'
  const lightTheme = highlighterConfig?.themeLight ?? 'vitesse-light'
  shikiHighlighterInstance.value = await createHighlighter({
    langs,
    themes: [darkTheme, lightTheme],
  })
  if (highlighterConfig.twoslash) {
    shikiHighlighterInstance.twoslashTransformer = createTransformerFactory(
      await createTwoslasher(highlighterConfig.twoslash === true
        ? {
            compilerOptions: {
              module: 199,
              moduleResolution: 99,
              jsx: 1,
              types: ['@sveltepress/vite/types', '@sveltepress/theme-default/types', '@sveltepress/theme-default/components', '@sveltejs/kit'],
            },
          }
        : highlighterConfig.twoslash),
    )(
      {
        langs: ['ts', 'tsx', 'svelte'],
        renderer: rendererFloatingSvelte(),
      },
    )
  }
}

export default highlighter
