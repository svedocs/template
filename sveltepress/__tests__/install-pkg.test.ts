import { mdToSvelte } from '@sveltepress/vite'
import { describe, expect, it } from 'vitest'
import highlighter, { initHighlighter } from '../src/markdown/highlighter'
import installPkg, { pkgRe } from '../src/markdown/install-pkg'

describe('install-pkg', async () => {
  await initHighlighter()

  it('re', () => {
    expect(pkgRe.exec('@install-pkg(foo-bar)')).toMatchInlineSnapshot(`
      [
        "@install-pkg(foo-bar)",
        "foo-bar",
        undefined,
      ]
    `)
    expect(pkgRe.exec('@install-pkg(foo-bar,create)')).toMatchInlineSnapshot(`
      [
        "@install-pkg(foo-bar,create)",
        "foo-bar,create",
        undefined,
      ]
    `)
  })

  it('md', async () => {
    const source = '@install-pkg(@sveltepress/vite,create)'

    const { code } = await mdToSvelte({
      mdContent: source,
      filename: 'install-pkg.md',
      highlighter,
      remarkPlugins: [installPkg],
    })
    expect(code).toMatchSnapshot()
  })
})
