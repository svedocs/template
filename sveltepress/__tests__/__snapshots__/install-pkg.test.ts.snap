// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`install-pkg > md 1`] = `
"<InstallPkg>{#snippet npm()}
<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    
    <!-- svelte-ignore a11y_no_noninteractive_tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"><span style="color:#59873A;font-style:inherit;--shiki-dark:#82AAFF;--shiki-dark-font-style:italic">npm</span><span style="color:#B56959;--shiki-dark:#ECC48D"> create</span><span style="color:#B56959;--shiki-dark:#ECC48D"> @sveltepress/vite</span></span></code></pre>
    <div class="svp-code-block--lang">
      sh
    </div>
    <CopyCode />
    
  </div>
</div>{/snippet}{#snippet yarn()}
<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    
    <!-- svelte-ignore a11y_no_noninteractive_tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"><span style="color:#59873A;font-style:inherit;--shiki-dark:#82AAFF;--shiki-dark-font-style:italic">yarn</span><span style="color:#B56959;--shiki-dark:#ECC48D"> create</span><span style="color:#B56959;--shiki-dark:#ECC48D"> @sveltepress/vite</span></span></code></pre>
    <div class="svp-code-block--lang">
      sh
    </div>
    <CopyCode />
    
  </div>
</div>{/snippet}{#snippet pnpm()}
<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    
    <!-- svelte-ignore a11y_no_noninteractive_tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"><span style="color:#59873A;font-style:inherit;--shiki-dark:#82AAFF;--shiki-dark-font-style:italic">pnpm</span><span style="color:#B56959;--shiki-dark:#ECC48D"> create</span><span style="color:#B56959;--shiki-dark:#ECC48D"> @sveltepress/vite</span></span></code></pre>
    <div class="svp-code-block--lang">
      sh
    </div>
    <CopyCode />
    
  </div>
</div>{/snippet}</InstallPkg>"
`;
